node/flatpak_node_generator/providers/npm.py

```diff
@@ -94,12 +94,74 @@ def _process_packages_v1(
             if 'dependencies' in info:
                 yield from self._process_packages_v1(lockfile, info)
 
+    def _process_packages_v2(
+        self, lockfile: Path, entry: Dict[str, Dict[Any, Any]]
+    ) -> Iterator[Package]:
+        for install_path, info in entry.get('packages', {}).items():
+            if (info.get('dev') or info.get('devOptional')) and self.no_devel:
+                continue
+            if info.get('link'):
+                # NOTE We're not interested in symlinks, NPM will create them at install time
+                # but we still could collect package symlinks anyway just for completeness
+                continue
+
+            name = info.get('name')
+
+            source: PackageSource
+            package_json_path = lockfile.parent / install_path / 'package.json'
+            if package_json_path.exists():
+                source = LocalSource(path=install_path)
+                if not name:
+                    with package_json_path.open('rb') as fp:
+                        name = json.load(fp)['name']
+            elif 'resolved' in info:
+                resolved_url = urllib.parse.urlparse(info['resolved'])
+                if resolved_url.scheme == 'file':
+                    source = LocalSource(path=resolved_url.path)
+                elif resolved_url.scheme in {'http', 'https'}:
+                    integrity = Integrity.parse(info['integrity'])
+                    # NOTE I don't know how to determine if the source came from a registry
```

Comment:
Wow, does npm not barf at the lack of cached metadata responses? If not that's actually pretty awesome, having to add that is really ugly.

Reply:
To my understanding, NPM doesn't need cached registry api responses as long as both `resolved` and `integrity` fields are present in the lockfile, but sadly it's not always the case.

Reply:
Ah and that falls back to the "idk" case right now.